File: app/src/main/java/com/deliveryhero/android/prototype/tabs/SectionItemsAdapter.java
package com.deliveryhero.android.prototype.tabs;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.ButterKnife;
import com.deliveryhero.android.prototype.R;
import com.deliveryhero.android.prototype.model.presentation.Section;

/**
 * Adapter implementation used to display the list of items under a section. This is intended as a
 * sample adapter implementation and should not be treated as production-ready.
 */
class SectionItemsAdapter extends RecyclerView.Adapter<SectionItemsAdapter.Holder> {

  private static final int DESCRIPTION = 0xf00;
  private static final int ITEM = 0xba0;

  private final LayoutInflater mInflater;
  private final Section mSection;

  SectionItemsAdapter(Context context, Section section) {
    mInflater = LayoutInflater.from(context);
    mSection = section;
  }

  @Override public int getItemViewType(int position) {
    return position == 0 ? DESCRIPTION : ITEM;
  }

  @Override public SectionItemsAdapter.Holder onCreateViewHolder(ViewGroup parent, int viewType) {
    int layout = viewType == DESCRIPTION ? R.layout.content_description : R.layout.content_item;
    return new SectionItemsAdapter.Holder(mInflater.inflate(layout, parent, false));
  }

  @Override public void onBindViewHolder(SectionItemsAdapter.Holder holder, int position) {
    holder.text.setText(getTextFor(position));
  }

  private String getTextFor(int position) {
    if (position == 0) {
      return mSection.getDescription();
    } else {
      return mSection.getItems().get(position - 1);
    }
  }

  @Override public int getItemCount() {
    return mSection.getItems().size() + 1;
  }

  /**
   * Simple view holder for the adapter
   */
  static class Holder extends RecyclerView.ViewHolder {

    @BindView(R.id.content_view) TextView text;

    Holder(View itemView) {
      super(itemView);
      ButterKnife.bind(this, itemView);
    }
  }
}

File: app/src/main/java/com/deliveryhero/android/prototype/tabs/SectionItemsAdapter.kt
package com.deliveryhero.android.prototype.tabs

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.deliveryhero.android.prototype.R
import com.deliveryhero.android.prototype.model.presentation.Section
import com.deliveryhero.android.prototype.tabs.SectionItemsAdapter.Holder
import kotlinx.android.synthetic.main.content_description.view.*

/**
 * Adapter implementation used to display the list of items under a section. This is intended as a
 * sample adapter implementation and should not be treated as production-ready.
 */
class SectionItemsAdapter(context: Context, private val section: Section) :
        RecyclerView.Adapter<Holder>() {

    private val inflater = LayoutInflater.from(context)

    override fun getItemViewType(position: Int) = if (position == 0) DESCRIPTION else ITEM

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
            Holder(inflater.inflate(layoutForType(viewType), parent, false))

    private fun layoutForType(viewType: Int) = when (viewType) {
        DESCRIPTION -> R.layout.content_description
        else -> R.layout.content_item
    }

    override fun onBindViewHolder(holder: Holder, position: Int) {
        holder.content.text = getTextFor(position)
    }

    private fun getTextFor(position: Int) = when (position) {
        0 -> section.description
        else -> section.items[position - 1]
    }

    override fun getItemCount() = section.items.size + 1

    class Holder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal val content: TextView = itemView.content_view
    }

    companion object {
        private val DESCRIPTION = 0xf00
        private val ITEM = 0xba0
    }
}